content/en/blog/2022/how-to-get-started-with-opentelemetry-go.md

```diff
@@ -0,0 +1,385 @@
+func main() {
+connectMongo()
+setupWebServer()
+}
+func connectMongo() {
+opts := options.Client()
+opts.ApplyURI("mongodb://localhost:27017")
+client, \_ = mongo.Connect(context.Background(), opts)
+//Seed the database with todo's
+docs := []interface{}{
+bson.D{{"id", "1"}, {"title", "Buy groceries"}},
+bson.D{{"id", "2"}, {"title", "install Aspecto.io"}},
+bson.D{{"id", "3"}, {"title", "Buy dogz.io domain"}},
+}
+client.Database("todo").Collection("todos").InsertMany(context.Background(), docs)
+}
+func setupWebServer() {
+r := gin.Default()
+r.GET("/todo", func(c *gin.Context) {
+collection := client.Database("todo").Collection("todos")
+//Important: Make sure to pass c.Request.Context() as the context and not c itself - TBD
+cur, findErr := collection.Find(c.Request.Context(), bson.D{})
+if findErr != nil {
+c.AbortWithError(500, findErr)
+return
+}
+results := make([]interface{}, 0)
+curErr := cur.All(c, &results)
+if curErr != nil {
+c.AbortWithError(500, curErr)
+return
+}
+c.JSON(http.StatusOK, results)
+})
+\_ = r.Run(":8080")
+}
+```
+
+Now that our small todo app is ready, let’s introduce OpenTelemetry.
+
+### Step 2: Install OpenTelemetry Go
+
+We will be configuring OpenTelemetry to instrument our Go app.
+
+1. To install the OTel SDK, run:
+
+```go
+go get go.opentelemetry.io/otel /
+go.opentelemetry.io/otel/sdk /
+```
+
+2. Instrument our Gin and Mongo libraries to generate traces.
+
+3. Gin instrumentation: Install otelgin
+
+```go
+go get go.opentelemetry.io/contrib/instrumentation/github.com/gin-gonic/gin/otelgin
+```
+
+4. Mongo instrumentation: Install otelmongo
+
+```go
+go get go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo
+```
+
+### Gin instrumentation: gin.Context
+
+We previously discussed the idea of context propagation – the way to transfer metadata between distributed services to correlate events in our system.
+
+The Gin framework has its own type gin.Context which gets passed as a parameter to an HTTP handler. However, the context that should be passed down to the mongo operations is the standard Go library Context object, available in gin.Context.Request.Context.
+
+```go
+//Make sure to pass c.Request.Context() as the context and not c itself
+cur, findErr := collection.Find(c.Request.Context(), bson.D{})
+So make sure that you pass the Context to the mongodb operation. Check out this issue for more info.
+```
+
+We now have our todo app ready and instrumented. It’s time to utilize OpenTelemetry to its full potential. Our ability to visualize traces is where the true troubleshooting power of this technology comes into play.
+
+For visualization, we’ll be using the open-source Jaeger Tracing and Aspecto.
+
+## Visualization with Jaeger
+
+The setup to export traces to Jaeger or Aspecto is relatively similar. Follow along with the Jaeger setup, then switch to Aspecto by changing a few lines of code.
+
+### OpenTelemetry Go and Jaeger Tracing: Export traces to Jaeger
```

Comment:
Now that Jaeger supports OTLP natively, should this section change to just set up the OTLP exporter and then send to jaeger that way?

Reply:
Thats a great idea, we will update the blog post in the future =)